### ISSUE TITLE
UnicodeDecodeError when rendering Unicode string in pillar

### ISSUE THREAD

**Issue by optix2000** [saltstack/salt#32975]
### Jinja throws a UnicodeDecode error when rendering a Unicode string from a pillar

If there's Unicode in the jinja file itself, the exception handler in salt.util.templates will throw the same exception trying to print out log messages.
#### States

test.sls:

```
test:
  file.managed:
    - name: /tmp/test.salt
    - template: jinja
    - source: salt://test.jinja
```

test.jinja:

```
{{ pillar.test }}
```

test2.sls

```
test:
  file.managed:
    - name: /tmp/test.salt
    - template: jinja
    - source: salt://test2.jinja
```

test2.jinja:

```
{{ pillar.test }}
なのです
```
#### Pillars

top.sls:

```
base:
    '*':
        - test
```

test.sls:

```
test: なのです
```
#### Case 1
1. Set up test state and pillar
2. `sudo salt-call -l debug state.sls test`

```
[ERROR   ] Unable to manage file: Jinja error: 'ascii' codec can't decode byte 0xe3 in p osition 0: ordinal not in range(128)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 366, in render_j inja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 754, in handle_exc eption
    reraise(exc_type, exc_value, tb)
  File "<template>", line 1, in top-level template code
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 0: ordinal not in r ange(128)

; line 1

---
{{ pillar.test }}    <======================

Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 366, in render_j inja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 989, in render
[...]
```
#### Case 2
1. Set up state and pillar
2. `sudo salt-call -l debug state.sls test2`

```
[ERROR   ] Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 178, in render_tmpl
    output = render_str(tmplstr, context, tmplpath)
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 408, in render_jinja_tmpl
    log.debug("TmplStr: {0}".format(tmplstr))
UnicodeEncodeError: 'ascii' codec can't encode characters in position 18-21: ordinal not in range(128)
```
### Versions Report

```
Salt Version:
           Salt: 2015.8.8

Dependency Versions:
         Jinja2: 2.8
       M2Crypto: Not Installed
           Mako: 1.0.4
         PyYAML: 3.11
          PyZMQ: 15.1.0
         Python: 2.7.11+ (default, Apr 17 2016, 14:00:29)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: 1.3.7
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: 0.3.8
          smmap: 0.9.0
        timelib: Not Installed

System Versions:
           dist: debian stretch/sid
        machine: x86_64
        release: 4.5.0-1-amd64
         system: debian stretch/sid

```

**Comment by jfindlay** [comment]
@optix2000, thanks for reporting.  The first case is an upstream bug with jinja, but the second should be easy to fix.
```py
from salt.utils.locales import sdecode
...
log.debug(u'TmplStr: {0}'.format(sdecode(tmplstr)))
```

**Comment by optix2000** [comment]
Could the first case be fixed by having pillars render into Unicode strings instead instead of byte arrays?

**Comment by jfindlay** [comment]
@optix2000, I don't know what the reason for different encoding choices would be, so I am not sure.  Ping @s0undt3ch.

**Comment by s0undt3ch** [comment]
Does this work?

```
log.debug('TmplStr: %s', tmplstr)
```

**Comment by damon-atkins** [comment]
It would be nice if log.* check to see if a string is unicode or not and then automatically added the decoding  e.g. decode("unicode_escape").

**Comment by s0undt3ch** [comment]
Decoding is not that simple but the above call that I provided does do some footwork to make it all work. Does it?

**Comment by damon-atkins** [comment]
unicode_escape takes unicodes and converts them to code points i.e. the string will only contain 0 to 126 everything above 126 will be escaped. 
>>> import sys
>>> sys.getdefaultencoding()
'ascii'
>>> import locale
>>> locale.getdefaultlocale()
('en_AU', 'cp1252')
>>>

Basically log needs to target sys.getdefaultencoding() when outputting to the screen/console

**Comment by damon-atkins** [comment]
```
>>> from __future__ import unicode_literals
>>> UNICODETEST_WITH_SIGNS = 'Testing Unicode \N{COPYRIGHT SIGN},\N{TRADE MARK SIGN},\N{REGISTERED SIGN}'
>>> print UNICODETEST_WITH_SIGNS
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\salt\bin\lib\encodings\cp850.py", line 12, in encode
    return codecs.charmap_encode(input,errors,encoding_map)
UnicodeEncodeError: 'charmap' codec can't encode character u'\u2122' in position 18: character maps to <undefined>
>>> print UNICODETEST_WITH_SIGNS.encode("unicode_escape")
Testing Unicode \xa9,\u2122,\xae
```

**Comment by s0undt3ch** [comment]
get default encoding is reset to ascii once the python interpreter starts and we need to know the encoding the system supports...

**Comment by s0undt3ch** [comment]
See `salt/__init__.py`, `__define_global_system_encoding_variable__()` for more details...

**Comment by optix2000** [comment]
I don't believe this is an issue decoding/encoding Unicode, but rather an issue of wrong types being passed around.

Based on the [Jinja docs](http://jinja.pocoo.org/docs/dev/api/#unicode), Jinja either accepts Unicode objects, or byte arrays (strings) with only ASCII characters (0-128). Based on the error messages, it appears that the pillar is being passed as a byte array of Unicode characters, rather than a Unicode object.

After playing around with PyYAML, it looks like PyYAML reads in Unicode strings as Unicode objects. This means that between the pillar data gets loaded from the YAML file and when it gets put into the Jinja, it becomes a byte array instead. This also seems related to #3436

This is mirrored by the [Python 2 documentation for Unicode](https://docs.python.org/2/howto/unicode.html), where it mentions that if byte arrays are passed to Unicode functions without an encoding type, they'll be assumed to be ASCII and thus throw `UnicodeDecodeError: 'ascii' codec can't decode byte`, which seems to be what's occurring.

**Comment by xofyarg** [comment]
I can confirm this problem on 2016.3.3. The fix in #3436 changes part of the context in template, which means `pillar.get('foo')` is successfully decoded, but `salt['pillar.get']('foo')` is not. I'm trying to make the decode function to be called a bit earlier, in pillar receiving stage.

The PR is still in progress since it has fix only for `RemotePillar`, not `AsyncRemotePillar`. I'm not sure if it's the right place to do this, please confirm, Thanks.

**Comment by damon-atkins** [comment]
Something I posted to the PY3 Issues
from __future__ import unicode_literals or from __future__ import absolute_import, **unicode_literals**
I have been slowing adding this to windows, since much of Windows is Unicode
It needs to be added to all *.py files to help with porting. Can it be added post Carbon Release RC1 to develop?
FYI https://docs.djangoproject.com/en/1.10/topics/python3/

**Comment by F30** [comment]
FWIW, I worked around this by using `{{ pillar['...'].decode('utf-8') }}` in the template. It's not very nice, as you have to know where in the Pillar non-ASCII data is stored, but works.

Additionally, the Salt output now says "Replace text file with binary file" instead of a diff, which is comprehensible but still wrong for UTF-8 data.

**Comment by damon-atkins** [comment]
Hi @F30  you might be interested in this function
`def _yaml_result_unicode_to_utf8(data):` within  salt/renderers/yaml.py`  forces everything to a PY2 string (which would be bytes in PY3)

Try on a test box, as it looks like this function has been removed in development.
```
def _yaml_result_unicode_to_utf8(data):
  return data
```

Also see https://github.com/saltstack/salt/issues/42278#issuecomment-318913952 `salt logging with plain str (ascii) with unicode chars causes UnicodeDecodeError PY2`   which contains a fix for logging which forces PY2 strings into PY2 Unicode when they can not be logged.

**Comment by F30** [comment]
@damon-atkins Sorry for the delay, I have now taken a look at `_yaml_result_unicode_to_utf8()` and I'm not really sure what to gain from it: As far as I can tell, it was only used internally to do the actual decoding when "yaml_utf8" was enabled in the config. I can see no way to use it manually from a Jinja template. If I recall correctly, I even tried enabling the "yaml_utf8" setting and it didn't solve my issue.

Looking at the code, this comes at no surprise, since what the function did was **en**coding a Unicode string to a UTF-8 binary string. However, in my problem the data already appears to be a UTF-8 binary string (even without "yaml_utf8", probably these are just the raw bytes from the UTF-8-encoded YAML file), which has to be **de**coded to a Unicode string in order to be accepted by Jinja. Commit 3184168365, which you're referencing, did some major changes to encoding and appears to also have removed the "yaml_utf8" setting (which is why `_yaml_result_unicode_to_utf8()` is not needed anymore).

As for the logging, I guess you might be right and my solution maybe could cause issues in higher log levels. I haven't considered that.

**Comment by terminalmage** [comment]
This is fixed for the upcoming 2018.3.0 release as part of a greater effort toward improving unicode support. This can be confirmed using the below Docker commands (run from the root of a git clone of Salt).

```
% docker run --rm -it -v $PWD:/testing terminalmage/issues:32975 sh -c 'salt-call state.apply test; cat /tmp/test.salt'
local:
----------
          ID: test
    Function: file.managed
        Name: /tmp/test.salt
      Result: True
     Comment: File /tmp/test.salt updated
     Started: 16:45:29.328105
    Duration: 55.502 ms
     Changes:
              ----------
              diff:
                  New file
              mode:
                  0644

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  55.502 ms
なのです
% docker run --rm -it -v $PWD:/testing terminalmage/issues:32975 sh -c 'salt-call state.apply test2; cat /tmp/test.salt'
local:
----------
          ID: test
    Function: file.managed
        Name: /tmp/test.salt
      Result: True
     Comment: File /tmp/test.salt updated
     Started: 16:45:55.414350
    Duration: 55.625 ms
     Changes:
              ----------
              diff:
                  New file
              mode:
                  0644

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  55.625 ms
なのです
なのです
```

You'll want to make sure you check out the 2018.3 branch after cloning the repo:

```bash
git clone https://github.com/saltstack/salt
cd salt
git checkout -b 2018.3 origin/2018.3
```

You can then proceed to run the docker commands above.
